tests/test_read_data.py

```diff
@@ -0,0 +1,544 @@
+        read_kwargs = {"blocksize": "128Mib"}
+
+    df = func(
+        input_files=input_files,
+        backend=backend,
+        file_type=file_type,
+        add_filename=add_filename,
+        input_meta=None,
+        columns=list(cols_to_select) if cols_to_select else None,
+        **read_kwargs,
+    )
+    if not cols_to_select:
+        cols_to_select = ["id", "text"]
+
+    assert list(df.columns) == list(df.head().columns)
+    if not add_filename:
+        # assert list(df.columns) == sorted(cols_to_select)
+        assert list(df.columns) == sorted(cols_to_select)
+    else:
+        # assert list(df.columns) == sorted(cols_to_select + ["filename"])
+        assert list(df.columns) == sorted(cols_to_select + ["filename"])
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+@pytest.mark.parametrize(
+    "function_name", ["read_data_blocksize", "read_data_files_per_partition"]
+)
+@pytest.mark.parametrize(
+    "input_meta", [{"id": "str"}, {"text": "str"}, {"id": "str", "text": "str"}]
+)
+def test_read_data_input_meta(
+    mock_multiple_jsonl_files, backend, function_name, input_meta
+):
+    if function_name == "read_data_files_per_partition":
+        func = read_data_files_per_partition
+        read_kwargs = {"files_per_partition": 1}
+    elif function_name == "read_data_blocksize":
+        func = read_data_blocksize
+        read_kwargs = {"blocksize": "128Mib"}
+
+    df = func(
+        input_files=mock_multiple_jsonl_files,
+        backend=backend,
+        file_type="jsonl",
+        add_filename=False,
+        input_meta=input_meta,
+        columns=None,
+        **read_kwargs,
+    )
+
+    if function_name == "read_data_files_per_partition" and backend == "cudf":
+        assert list(df.columns) == list(input_meta.keys())
+    else:
+        # In the read_data_fpp case, because pandas doesn't support `prune_columns`, it'll always return all columns even if input_meta is specified
+        # In the `read_data_blocksize` case, `dask.read_json` also doesn't `prune_columns` so it'll always return all columns
+        # if you user wants to select subset of columns, they should use `columns` parameter
+        assert list(df.columns) == ["id", "text"]
+
+
+@pytest.mark.parametrize(
+    "backend",
+    [
+        "pandas",
+        pytest.param("cudf", marks=pytest.mark.gpu),
+    ],
+)
+@pytest.mark.parametrize("file_type", ["parquet", "jsonl"])
+@pytest.mark.parametrize(
+    "read_kwargs",
+    [
+        *[({"files_per_partition": fpp, "blocksize": None}) for fpp in range(1, 6)],
+        *[
+            ({"blocksize": bs, "files_per_partition": None})
+            for bs in ["128MiB", "256MiB", "512MiB"]
+        ],
+    ],
+)
+def test_read_data_different_columns(
+    mock_multiple_jsonl_files_different_cols,
+    mock_multiple_parquet_files_different_cols,
+    backend,
+    file_type,
+    read_kwargs,
+):
```

Comment:
In this test case we try to read json / parquet directories which have some files containing `id / text` columns while some containing `text / meta / id` columns (we try to mimic RPV1 dataset). 

We try reading with both APIs of blocksize and FPP. 

Our goal is to read only `id` and `text` fields since those are common across all files. However we notice that 3/4 of the blocksize test cases ie `read_kwargs-5/6/7` fail (except when backend=='pandas' and filetype=='parquet') and I suspect that's because `dd.read_json` / `dd.read_parquet` doesn't nicely seem to support mismatched schemas.

Would be amazing if I can get a third eye here since that's the only case left for this PR. Worst case we can mark it as `xfail` until `dd` can support mismatched schemas (in which case for RPV1 like datasets we'll have to go down the RPV1 path :( )